### ISSUE TITLE
Bootstrap nodes seems blocked by GFW (China)

### ISSUE THREAD

**Issue by blurHY** [ipfs/go-ipfs#5993]
There is no peers connected before executing `ipfs swarm connect ...`

**Comment by hoogw** [comment]
GFW, is china government build great firewall, has a white list. 
Any IP does not on their white list will be blocked.

Only a few IP on white list is allow access, otherwise NOT on white list, will be blocked. 
Your nodes not on china government's white list, so blocked.

**Comment by Stebalien** [comment]
Last time I checked, GFW was blacklist based. The solution here is probably to try to remember nodes that are reachable between restarts so we can bootstrap off of them. In this case, those would be other nodes in China.

**Comment by hoogw** [comment]
Check here for how to by pass china's great firewall.


https://github.com/breakwa11/gfw_whitelist

https://github.com/pexcn-legacy/gfw-whitelist

https://github.com/neko-dev/gfw_whitelist

https://github.com/pexcn/daily

**Comment by blurHY** [comment]
Nope, GFW is blacklist based. 
Could IPFS remember nodes like i2p?

**Comment by blurHY** [comment]
Gov's hackers have already add IPFS's default bootstrap nodes to blacklist. So without `ipfs swarm connect` to add a non-blocked peer, Chinese users will NOT connect to any peers

**Comment by Mikaela** [comment]
>  Could IPFS remember nodes like i2p?

https://github.com/ipfs/go-ipfs/issues/3926 and I think this issue may be a duplicate of https://github.com/ipfs/go-ipfs/issues/3908.

**Comment by blurHY** [comment]
Duplicate

**Comment by hoogw** [comment]
Chinese government block chinese user connect to each other.  How to work around it?

**Comment by Stebalien** [comment]
Reopened because this issue is useful to track IPFS use in China.

@hoogw within china? That is, you're having trouble connecting two nodes inside china to each other?

**Comment by blurHY** [comment]
So far, GFW could only block bootstrap nodes.

**Comment by LetItGlow** [comment]
When I run my daemon, I get a lot of China nodes, how do they even connect to me?

**Comment by blurHY** [comment]
Maybe they can't. Because GFW can detect IPFS traffic and block it. 
https://github.com/HelloZeroNet/ZeroNet/issues/1914

**Comment by LetItGlow** [comment]
Any ways to encrypt or 'key' the traffic IPFS side so it isn't obvious for packet sniffers (or whatever they are using) anymore?

**Comment by Stebalien** [comment]
> Any ways to encrypt or 'key' the traffic IPFS side so it isn't obvious for packet sniffers (or whatever they are using) anymore?

We already do.

Our current issues are:

1. We usually run on port 4001 (obvious). We've considered listening on a random port by default, it just makes port forwarding a bit trickier. However, we need to seriously revisit this now.
2. We negotiate our crypto protocol in the clear and, worse, use a custom TLS-like (but not TLS) protocol. We're working on switching to plain TLS which should make IPFS nodes look like normal web traffic.

Unfortunately, even if we fix all this, it's still pretty trivial to connect to a computer and _check_ if it's an IPFS node.

**Comment by LetItGlow** [comment]
Why don't we just ask nicely, after we exchanged keys?
So we go on with the usual Client and server hello, exchange SSL "business cards" and then just ask bluntly: "Are you IPFS, gimme your ID"
If the server doesn't understand, just leave the room.

**Comment by Stebalien** [comment]
That's effectively how the TLS security transport will work. However, the firewall will still be able to politely ask, discover all computers running IPFS, and block them.

**Comment by izern** [comment]
@LetItGlow 
/ip4/39.106.19.168/tcp/4001/ipfs/QmX5KX7nA3PwhCgSswaxavGmNvGKP9un6rU5VzXP8Mduyy

**Comment by blurHY** [comment]
Any progress ?

**Comment by Stebalien** [comment]
No. Unfortunately, the best solution for the moment is to add an alternative bootstrapper by running something like (to use @izern's node):

```sh
ipfs bootstrap add   /ip4/39.106.19.168/tcp/4001/ipfs/QmX5KX7nA3PwhCgSswaxavGmNvGKP9un6rU5VzXP8Mduyy
```
